### DISCUSSION TITLE
Add possibility to omit calling Reconcile methods per tick

### DISCUSSION THREAD

**Discussion by tedbarth** [FirstGearGames/FishNet#283] [Feature Request]
FishNet's CSP using Reconcile and Replicate methods can't well work with on-demand updates, meaning that you always have to call the Replicate method with some data per tick **on the client**. But not calling the Replicate method is something you want to do to safe bandwidth, especially if the Replicate data struct has some more properties.
If you instead just omit the call of the Replicate method on the client, the Replicate method on the server is called with a default replicate data object as argument, which in most cases is no valid data.

Additionally there are use cases where objects (e.g. projectiles) may not want to send user input data at all to the server but still want to receive more complex updates from the server and benefit from **PredictedObject** implementation (Interpolation, etc.). In those situations **NetworkTransform** just doesn't send enough data. Other frameworks have additional components, for instance **NetworkRigidBody**. But even those might not be enough. In such situation (e.g. when wanting to send transform, rigidbody and other data - Only those properties you really want to sync - and don't want to send actions) the only way to always receive valid Replicate data in the Replicate method body on the server is to send an empty Struct. This probably is not bandwidth-friendly.

**Proposal:**
- Add stability to code (no idea if there is anything to do) so that not sending Replicate data is safe
- Don't invoke the body of the Reconcile method if no data is present, transparently

**Comment by FirstGearGames** [comment]
Reconcile doesn't invoke if there's no data.
Server only sends reconciles if there is change to send. But you don't have to send every tick anyway. On the server just do something like this....
```cs
if (base.TimeManager.LocalTick % 3 == 0) Reconcile(...) //asServer true.
```
If the client doesn't have a reconcile for the tick nothing bad will happen it will just catch the next one.

**Comment by tedbarth** [comment]
It is not about not sending Reconcile data to the client. Instead I want to avoid sending Replicate data to the server.

With FN 3.4.3 (and all 3.x.x I tested earlier) the server still calls my Reconcile method's body (below `private void Move (...)`) per tick, even if the client didn't call it locally (as part of HandleAxesInput here).

```
  protected override void OnNetworkTick() {
    if (IsOwner) {
      Reconcile(default, false);

      // Where input is read and values for the forces to apply are calculated (no forces are applied here)
      //HandleAxesInput(); // Calls [Reconcile] Move() with filled MoveData internally
    }

    if (IsServer) {
      Move(default, true);
    }
  }
```

![image](https://user-images.githubusercontent.com/8289532/223639898-fe1e6387-b569-4897-a67e-41271194e1c3.png)

MoveData here matches MoveData's `default` state.

**Comment by FirstGearGames** [comment]
Both replicate and reconcile methods will only call, if you actually call them in your code. If you do not have any lines calling the methods, they will not run.

However, replicates are intentionally run(assuming you call them) even when data is default. This is important because you might want to do something like add gravity in the replicate. But on the plus side, unless you are actually changing inputs the data will not be sent to the server. Keep in mind, there is some redundancy but eventually it stops sending.

**Reply by tedbarth** [reply]
Then we have a bug instead: My Replicate method `Move` is definitively called on the server when not called on the client, every tick, as you can see on the screenshot. There are only two references to `Move(...)`, one in `OnNetworkTick()` secured by `IsServer` and one as part of the explicitly not called `HandleAxesInput()` secured by `IsOwner`. `HandleAxesInput()` also is not called anywhere else but in the disabled call.

**Reply by tedbarth** [reply]
I set a break point on `MoveData::SetTick(uint value)` which never gets called on client and server.

Then I looked at the [caller](https://github.com/FirstGearGames/FishNet/blob/main/Assets/FishNet/Runtime/Object/NetworkBehaviour.Prediction.cs#L452) of my `Move()` method, when a `default` data was passed in. I am now even more confused, not understanding why you explicitly call the replicate method with `default` data in this method on every frame, when the client didn't sent any data in this tick:
```
        /// <summary>
        /// Gets the next replicate in queue.
        /// </summary>
        [CodegenMakePublic] //internal
        [APIExclude]
        public void Replicate_Server_Internal<T>(ReplicateUserLogicDelegate<T> del, Queue<T> q, Channel channel) where T : IReplicateData
        {
            int count = q.Count;
            if (count > 0)
            {
                ....
            }
            else
            {
                del.Invoke(default, true, channel, false); // HERE
            }
        }
```
IMHO this should never be the case unless explicitly called like this *on the client*. Could you elaborate on why you do this?

**Reply by FirstGearGames** [reply]
It's called every tick, not frame. But it's done because you still may want to perform actions even when there is no input, such as gravity.

**Reply by tedbarth** [reply]
But why then is the tick set to default (0), too?
What is the best way to identify those cases? `tick == 0` is a bit weird, API wise.
Can packages that are dropped/delayed on the network also lead to such "default call"?
Shouldn't this be mentioned in the docs somewhere in the CSP tutorial, maybe as part of the final script code?

**Reply by Fdrph** [reply]
I found this out the hard way too, i was receiving these default structs with input set to 0 which was resetting my accel code because it thought the input was legit and the joystick was on (0,0). Checking if tick == 0 didn't fix it cuz i was getting moveData's with valid input data but with tick set to 0... I dunno man.
To differentiate between a valid moveData and one where: 

- A - because of packet loss this tick we don't have anything. 
- B - because nothing changed on the client nothing is sent.

I had to add a boolean on the client to moveData that is always set to true. in move() I then do: 
```
if (!md.IsDataPacket)
 {
    _noPacketAvailable = true;
    inputDir = _LastInputDir;
} else { _noPacketAvailable = false; }
```


this is hacky but works, problem is it still doesn't differentiate between the 3 possible states.
I suggest the client should send a lil boolean packet when nothing has changed: NoChange = true
Then on the server, we can know easily what to send to move() for the user:

- If we don't have moveData for the user because of packetloss (we detected a gap in the tick numbers in the packets) -> set boolean in move() parameters lostPacket = true this let's the user use input from last tick and assume it's continuing (which is what i do)
- If we don't have moveData for the user because input hasn't changed set inside moveData a lil boolean -> noChange = true

Then let the user use this info how they see fit and for their game's needs.

**Comment by FirstGearGames** [comment]
Added as gotcha/limitation inside guides. In prediction v2 this is much less of a concern/problem. Expect to see better options in version 2.
https://fish-networking.gitbook.io/docs/manual/guides/client-side-prediction/version-1/more/limitations-and-gotchas
